### ISSUE TITLE
Add JAX profiling infrastructure and profiling script

### ISSUE THREAD

**Issue by zeinali72** [zeinali72/swe-pinn#104]
## Context

The training pipeline has **zero profiling tooling**. There is no way to measure where wall-clock time is spent per epoch, how much GPU memory is used, or whether JIT recompilation is happening. This is the prerequisite for all other profiling work.

## Related Issues

- **Parallel-safe** with: all HPO issues (#88-#103)
- **Enables**: PDE Jacobian profiling, memory profiling, data transfer profiling, architecture benchmarks (filed separately)
- **Related**: #62 (zero tests — similar "missing infrastructure" theme)

## Current State

- No profiling tools in \`pyproject.toml\` or \`requirements.txt\`
- No JAX profiler configuration
- No TensorBoard integration
- \`src/monitoring/aim_tracker.py\` logs metrics but not compute timing
- \`src/monitoring/console_logger.py\` prints losses, no timing breakdown

### 1. Add profiling dependencies

\`\`\`toml
# pyproject.toml
[project.optional-dependencies]
profile = [
    "tensorboard>=2.14",
    "tensorboard-plugin-profile>=2.14",
    "jax[cuda12]",  # ensures jax.profiler available
]
\`\`\`

### 2. Create \`scripts/profile_training.py\`

A standalone script that runs a short training session with profiling enabled:

\`\`\`python
#!/usr/bin/env python3
"""Profile a training run and output JAX/TensorBoard traces.

Usage:
    python scripts/profile_training.py --config configs/experiment_1_mlp.yaml --epochs 10 --output profile_output/

Produces:
    profile_output/
    ├── jax_trace/              # JAX XLA trace (view with TensorBoard)
    ├── timing_breakdown.json   # Per-component wall-clock times
    └── memory_snapshot.txt     # Peak memory per device
"""
import argparse
import time
import json
import jax
import jax.numpy as jnp

def profile_epoch_components(config_path, n_epochs=10, output_dir="profile_output"):
    """Profile individual training components separately."""
    # 1. Setup (same as experiment train.py)
    # ...
    
    timings = {
        "jit_compilation": {"first_epoch": 0, "note": "Includes tracing + XLA compile"},
        "data_generation": {"per_epoch_ms": []},
        "forward_pass": {"per_epoch_ms": []},
        "pde_jacobian": {"per_epoch_ms": []},
        "backward_pass": {"per_epoch_ms": []},
        "optimizer_update": {"per_epoch_ms": []},
        "validation": {"per_epoch_ms": []},
        "host_device_transfer": {"per_epoch_ms": []},
    }
    
    for epoch in range(n_epochs):
        # Time each component separately
        t0 = time.perf_counter()
        scan_inputs = generate_epoch_data_jit(epoch_key)
        jax.block_until_ready(scan_inputs)  # Force sync
        timings["data_generation"]["per_epoch_ms"].append((time.perf_counter() - t0) * 1000)
        
        # ... similar for each component
        
        # On epoch 2, enable JAX trace (skip epoch 1 which includes JIT)
        if epoch == 2:
            jax.profiler.start_trace(os.path.join(output_dir, "jax_trace"))
        if epoch == 3:
            jax.profiler.stop_trace()
    
    # Memory snapshot
    memory_stats = {}
    for device in jax.devices():
        stats = device.memory_stats()
        if stats:
            memory_stats[str(device)] = {
                "peak_bytes": stats.get("peak_bytes_in_use", "N/A"),
                "current_bytes": stats.get("bytes_in_use", "N/A"),
            }
    
    # Save results
    with open(os.path.join(output_dir, "timing_breakdown.json"), "w") as f:
        json.dump(timings, f, indent=2)
    
    # Print summary
    print("\n=== TIMING BREAKDOWN (per epoch, ms) ===")
    for component, data in timings.items():
        if "per_epoch_ms" in data and data["per_epoch_ms"]:
            vals = data["per_epoch_ms"]
            print(f"  {component:<25}: mean={sum(vals)/len(vals):.1f}ms  "
                  f"min={min(vals):.1f}ms  max={max(vals):.1f}ms")
\`\`\`

### 3. Add timing hooks to \`src/training/loop.py\` (optional, behind flag)

\`\`\`python
# In run_training_loop(), controlled by config flag
enable_profiling = cfg.get("debug", {}).get("profile_timing", False)

if enable_profiling:
    t_data = time.perf_counter()
    
scan_inputs = generate_epoch_data_jit(epoch_key)

if enable_profiling:
    jax.block_until_ready(scan_inputs)
    timing["data_gen_ms"] = (time.perf_counter() - t_data) * 1000
\`\`\`

### 4. Add GPU memory monitoring utility

\`\`\`python
# src/utils/profiling.py
import jax

def log_memory_usage(label=""):
    """Print current GPU memory usage."""
    for device in jax.devices():
        stats = device.memory_stats()
        if stats:
            used_mb = stats.get("bytes_in_use", 0) / 1e6
            peak_mb = stats.get("peak_bytes_in_use", 0) / 1e6
            print(f"[{label}] {device}: {used_mb:.0f} MB used, {peak_mb:.0f} MB peak")

def check_jit_recompilation(fn, *args, **kwargs):
    """Run a function twice and warn if JIT recompiles on second call."""
    # First call: compiles
    t0 = time.perf_counter()
    result1 = fn(*args, **kwargs)
    jax.block_until_ready(result1)
    t1 = time.perf_counter()
    
    # Second call: should be cached
    result2 = fn(*args, **kwargs)
    jax.block_until_ready(result2)
    t2 = time.perf_counter()
    
    compile_ms = (t1 - t0) * 1000
    cached_ms = (t2 - t1) * 1000
    ratio = compile_ms / max(cached_ms, 0.001)
    
    print(f"First call: {compile_ms:.1f}ms | Second call: {cached_ms:.1f}ms | Ratio: {ratio:.1f}x")
    if ratio < 2.0:
        print("  WARNING: Low ratio suggests recompilation on second call!")
\`\`\`

## Files to Create

- \`scripts/profile_training.py\` — standalone profiling script
- \`src/utils/profiling.py\` — profiling utility functions (optional)

## Files to Modify

- \`pyproject.toml\` — add \`profile\` optional dependency group
- \`src/training/loop.py\` — optional timing hooks behind \`debug.profile_timing\` flag

## Acceptance Criteria

- [ ] \`pip install -e ".[profile]"\` installs TensorBoard + profiling deps
- [ ] \`python scripts/profile_training.py --config <config> --epochs 10\` produces timing breakdown
- [ ] JAX trace viewable in TensorBoard (\`tensorboard --logdir profile_output/jax_trace\`)
- [ ] GPU memory stats captured and printed
- [ ] JIT recompilation detector works
